### ISSUE TITLE
Completed Jobs not cleaned up from HTCondorCE

### ISSUE THREAD

**Issue by maxfischer2781** [DIRACGrid/DIRAC#4935]
This issue is a follow-up on [GGUS#150403](https://ggus.eu/index.php?mode=ticket_info&ticket_id=150403).

We can observe that VOs using DIRAC (LHCb, Belle, enmr.eu) consistently have a high number of "completed jobs" in the HTCondorCE queue. This regularly amounts to 10x to 100x the number of idle/running jobs. Jobs are only removed by the CE's internal cleanup rules:
```
[root@htcondor-ce-2-kit ~]# condor_ce_history belleprd -match 5 -af RemoveReason
CE job removed by SYSTEM_PERIODIC_REMOVE due to being in the hold state for 24 hours.
CE job removed by SYSTEM_PERIODIC_REMOVE due to being in the hold state for 24 hours.
CE job removed by SYSTEM_PERIODIC_REMOVE due to being in the hold state for 24 hours.
CE job removed by SYSTEM_PERIODIC_REMOVE due to being in the hold state for 24 hours.
CE job removed by SYSTEM_PERIODIC_REMOVE due to being in the hold state for 24 hours.
```
This is an issue because such jobs leave metadata on the CE and count towards [Condor ``MAX_JOBS_PER_OWNER``](https://htcondor.readthedocs.io/en/v8_8/admin-manual/configuration-macros.html#condor-schedd-configuration-file-entries); this can overload the CE or prevent job submission.

Note that it is unclear to me where the issue originates. At a glance, ``HTCondorCEComputingElement.getJobOutput`` appears to properly transfer the output.

https://github.com/DIRACGrid/DIRAC/blob/70004435a4133cc0982396f17d0e4c854aaddacc/src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py#L408

**Comment by andresailer** [comment]
As @VladimirRomanovsky wrote in the GGus ticket, DIRAC doesn't call getJobOutput for all jobs. 

Can you clarify, is it an issue even when you have SYSTEM_PERIODIC_REMOVE on the server?

**Comment by maxfischer2781** [comment]
We currently have the default HTCondorCE settings, that is MAX_JOBS_PER_OWNER=100000 and a SYSTEM_PERIODIC_REMOVE for this case of 24h. So yes, as the ticket illustrates this is an issue.

To give some rough numbers, for 3k running jobs we see 90k completed jobs at the moment (per CE).

**Comment by maxfischer2781** [comment]
This is related to but not superseded by htcondor/htcondor-ce#331. Be aware that PR added a default to leave the jobs in the queue for *30 days*. We already observe the issue with the regular removal (due to expired proxy, not queue time) after *1 day*.

**Comment by andresailer** [comment]
I see, your server removes the jobs once the output files are transferred, right?

Are these jobs completed or on hold? Your example shows "hold", but you say completed in other places

If pilots are held they should be removed more or less immediately. Though this isn't called if they go from completed to held, I would think. (it is also possible to disable this, but I cannot see that configuration for the relevant VOs)
https://github.com/DIRACGrid/DIRAC/blob/70004435a4133cc0982396f17d0e4c854aaddacc/src/DIRAC/Resources/Computing/HTCondorCEComputingElement.py#L397-L399

I am not sure there is a logic that would call condor_rm on completed pilots. I think for other CE types (Cream, ARC), the pilots were always eventually removed by the CEs.

**Comment by maxfischer2781** [comment]
> I see, your server removes the jobs once the output files are transferred, right?

Yes, HTCondorCE is basically a HTCondor Schedd with remote submission. Completed jobs remain in the queue until they are fetched via ``condor_transfer_data``.

> Are these jobs completed or on hold? Your example shows "hold", but you say completed in other places

Since the proxy expires after the job completes, jobs transition to "held" after completion; this is why the default SYSTEM_PERIODIC_REMOVE triggers for them. So it is completed->held->removed. In the queue the largest bulk is "completed":
```
[root@htcondor-ce-2-kit ~]# condor_ce_q plhcb49 -total


-- Schedd: htcondor-ce-2-kit.gridka.de : <192.108.45.34:1519?... @ 02/01/21 13:24:13
Total for query: 91396 jobs; 88493 completed, 0 removed, 100 idle, 2753 running, 50 held, 0 suspended
Total for all users: 101768 jobs; 93743 completed, 0 removed, 768 idle, 6820 running, 437 held, 0 suspended
```

> I think for other CE types (Cream, ARC), the pilots were always eventually removed by the CEs.

HTCondorCE also eventually removes the jobs. However, the delay is large enough for considerable number of jobs to accumulate.

**Comment by fstagni** [comment]
Dear @maxfischer2781 ,
we discussed this issue in the developers meeting last week, minutes in https://indico.cern.ch/event/1001396/note/ if you are interested.

Bottom line is that:
- we don't want to impose to any DIRAC user to run a local `schedd` even if this solution is already possible, and actually used by at least CLIC.
- a proper solution involves downloading all pilot outputs, independently of the CE type (being it HTCondor, ARC, SSH (DIRAC), vac, etc) basically wherever we run pilots. This is a development for which we will have little time/priority.

In LHCb we have applied a submit string : `periodic_remove =  (JobStatus == 4) && (time() - EnteredCurrentStatus) > ( 24 * 3600)` (added to `/Resources/Computing/HTCondorCE`). I will wait a few days and then I will add to documentation or to the defaults.
